agent/pkg/handler/manage_agent.go

```diff
@@ -0,0 +1,96 @@
+package handler
+
+import (
+	"time"
+
+	"github.com/buoyantio/linkerd-buoyant/agent/pkg/api"
+	"github.com/buoyantio/linkerd-buoyant/agent/pkg/k8s"
+	pb "github.com/buoyantio/linkerd-buoyant/gen/bcloud"
+	log "github.com/sirupsen/logrus"
+)
+
+type ManageAgent struct {
+	api    *api.Client
+	k8s    *k8s.Client
+	log    *log.Entry
+	stopCh chan struct{}
+}
+
+// NewManageAgent instantiates a new agent commands handler
+func NewManageAgent(k8sClient *k8s.Client, apiClient *api.Client) *ManageAgent {
+	log := log.WithField("handler", "manage_agent")
+	log.Debug("initializing")
+
+	return &ManageAgent{
+		api:    apiClient,
+		k8s:    k8sClient,
+		log:    log,
+		stopCh: make(chan struct{}),
+	}
+}
+
+// Start initiates ManageAgent handler
+func (h *ManageAgent) Start() {
+	for {
+		select {
+		case <-h.stopCh:
+			return
+		default:
+			agentCommand, err := h.api.RecvAgentCommand()
+			if err != nil {
+				h.log.Errorf("Error from stream %s", err)
+				continue
+			}
+			switch command := agentCommand.Command.(type) {
+			case *pb.AgentCommand_GetProxyDiagnostics:
+				proxyDiagnostic := command.GetProxyDiagnostics
+				go h.handleProxyDiagnostics(proxyDiagnostic.PodName, proxyDiagnostic.PodNamespace, proxyDiagnostic.DiagnosticId)
+			}
+		}
+	}
```

Comment:
rather than blocking on `h.api.RecvAgentCommand()`, what if that method returned a channel we can receive commands on, that way as soon as a `ManageAgent.Stop()` is called, `ManageAgent` stops trying to receive commands?

going a step further (and related to my comment around replacing `api.Client.CloseAgentCommandStream()` with a more general `api.Client.Close()`), if we can ensure that calling `ManageAgent.Stop()` will stop it from trying to receive, we should be able to safely call `api.Client.Close()` from main, and do away with the `client_closed` boolean in `manageAgentStream`.

Reply:
That sounds a lot more elegant. However we still need to put some async code in a goroutine to fill this chan. I assume this would live in the `manage_agent_stream.go`